File: vl53l1x_wiringpi.py
#!/usr/bin/env python3
"""
VL53L1X Distance Sensor - WiringPi версия
Использует библиотеку wiringPi для работы с I2C
"""

import time
import sys
from datetime import datetime

try:
    import wiringpi
except ImportError:
    print("Ошибка: Библиотека wiringPi не установлена.")
    print("Установите её командами:")
    print("  pip install wiringpi")
    print("или")
    print("  sudo apt install python3-wiringpi")
    sys.exit(1)


class VL53L1XWiringPi:
    """Драйвер VL53L1X на основе WiringPi"""

    # I2C адрес датчика
    DEFAULT_ADDRESS = 0x29

    # Основные регистры VL53L1X (16-битные адреса)
    SOFT_RESET = 0x0000
    FIRMWARE_SYSTEM_STATUS = 0x0010
    IDENTIFICATION_MODEL_ID = 0x010F

    # Регистры управления
    SYSTEM_INTERRUPT_CLEAR = 0x0086
    SYSTEM_MODE_START = 0x0087
    SYSTEM_SEQUENCE_CONFIG = 0x0001

    # Регистры статуса
    GPIO_TIO_HV_STATUS = 0x0031
    RESULT_RANGE_STATUS = 0x0089
    RESULT_FINAL_CROSSTALK_CORRECTED_RANGE_MM_SD0 = 0x0096

    # Регистры конфигурации
    RANGE_CONFIG_VCSEL_PERIOD_A = 0x0060
    RANGE_CONFIG_VCSEL_PERIOD_B = 0x0063
    RANGE_CONFIG_TIMEOUT_MACROP_A = 0x005E
    RANGE_CONFIG_TIMEOUT_MACROP_B = 0x0061
    RANGE_CONFIG_VALID_PHASE_HIGH = 0x0069
    SYSTEM_INTERRUPT_CONFIG_GPIO = 0x0046

    # Настройки дальности
    SD_CONFIG_WOI_SD0 = 0x0078
    SD_CONFIG_WOI_SD1 = 0x007A
    SD_CONFIG_INITIAL_PHASE_SD0 = 0x007C
    SD_CONFIG_INITIAL_PHASE_SD1 = 0x007E

    def __init__(self, i2c_bus=1, address=DEFAULT_ADDRESS):
        """
        Инициализация датчика

        Args:
            i2c_bus: Номер шины I2C (обычно 1 для Raspberry Pi)
            address: I2C адрес датчика
        """
        self.address = address
        self.i2c_bus = i2c_bus
        self.i2c_handle = None

    def open(self):
        """Открытие I2C соединения через WiringPi"""
        try:
            # Инициализация WiringPi
            wiringpi.wiringPiSetup()

            # Открываем I2C устройство
            self.i2c_handle = wiringpi.wiringPiI2CSetup(self.address)

            if self.i2c_handle < 0:
                print(f"✗ Не удалось открыть I2C устройство на адресе 0x{self.address:02X}")
                return False

            print(f"✓ I2C устройство открыто (handle: {self.i2c_handle})")

            # Проверка ID датчика
            model_id = self.read_reg16(self.IDENTIFICATION_MODEL_ID)
            if model_id == 0xEACC:
                print(f"✓ Датчик VL53L1X обнаружен (ID: 0x{model_id:04X})")
                return True
            else:
                print(f"⚠ Неизвестный датчик (ID: 0x{model_id:04X})")
                return False

        except Exception as e:
            print(f"✗ Ошибка открытия I2C: {e}")
            return False

    def write_reg8(self, reg, value):
        """Запись 8-битного значения в 16-битный регистр"""
        try:
            # Для 16-битного адреса регистра
            # Сначала записываем старший байт адреса
            wiringpi.wiringPiI2CWriteReg8(self.i2c_handle, (reg >> 8) & 0xFF, reg & 0xFF)
            # Затем записываем данные
            wiringpi.wiringPiI2CWrite(self.i2c_handle, value)
        except Exception as e:
            pass

    def write_reg16(self, reg, value):
        """Запись 16-битного значения в 16-битный регистр"""
        try:
            # Записываем адрес
            wiringpi.wiringPiI2CWriteReg8(self.i2c_handle, (reg >> 8) & 0xFF, reg & 0xFF)
            # Записываем данные (big-endian)
            wiringpi.wiringPiI2CWrite(self.i2c_handle, (value >> 8) & 0xFF)
            wiringpi.wiringPiI2CWrite(self.i2c_handle, value & 0xFF)
        except Exception as e:
            pass

    def read_reg8(self, reg):
        """Чтение 8-битного значения из 16-битного регистра"""
        try:
            # Записываем адрес регистра
            wiringpi.wiringPiI2CWriteReg8(self.i2c_handle, (reg >> 8) & 0xFF, reg & 0xFF)
            # Читаем данные
            return wiringpi.wiringPiI2CRead(self.i2c_handle) & 0xFF
        except Exception as e:
            return 0

    def read_reg16(self, reg):
        """Чтение 16-битного значения из 16-битного регистра"""
        try:
            # Записываем адрес регистра
            wiringpi.wiringPiI2CWriteReg8(self.i2c_handle, (reg >> 8) & 0xFF, reg & 0xFF)
            # Читаем 2 байта
            high = wiringpi.wiringPiI2CRead(self.i2c_handle) & 0xFF
            low = wiringpi.wiringPiI2CRead(self.i2c_handle) & 0xFF
            return (high << 8) | low
        except Exception as e:
            return 0

    def init_sensor(self):
        """Инициализация датчика VL53L1X"""
        print("Инициализация датчика...")

        # Программный сброс
        self.write_reg8(self.SOFT_RESET, 0x00)
        time.sleep(0.01)
        self.write_reg8(self.SOFT_RESET, 0x01)
        time.sleep(0.01)

        # Ждем загрузки прошивки
        print("Ожидание загрузки прошивки...")
        for i in range(100):
            status = self.read_reg8(self.FIRMWARE_SYSTEM_STATUS)
            if status & 0x01:
                print(f"✓ Прошивка загружена (попытка {i+1})")
                break
            time.sleep(0.01)
        else:
            print("⚠ Таймаут загрузки прошивки")

        # Базовая конфигурация
        # Настройки VCSEL (лазера)
        self.write_reg8(self.RANGE_CONFIG_VCSEL_PERIOD_A, 0x09)
        self.write_reg8(self.RANGE_CONFIG_VCSEL_PERIOD_B, 0x0D)

        # Настройки валидности фазы
        self.write_reg8(self.RANGE_CONFIG_VALID_PHASE_HIGH, 0xC8)

        # Настройки окна детектирования
        self.write_reg16(self.SD_CONFIG_WOI_SD0, 0x0001)
        self.write_reg16(self.SD_CONFIG_WOI_SD1, 0x0001)
        self.write_reg8(self.SD_CONFIG_INITIAL_PHASE_SD0, 0x00)
        self.write_reg8(self.SD_CONFIG_INITIAL_PHASE_SD1, 0x00)

        # Timing budget (время измерения)
        self.write_reg16(self.RANGE_CONFIG_TIMEOUT_MACROP_A, 0x00D6)
        self.write_reg16(self.RANGE_CONFIG_TIMEOUT_MACROP_B, 0x00D6)

        # Настройка прерываний
        self.write_reg8(self.SYSTEM_INTERRUPT_CONFIG_GPIO, 0x01)

        # Очистка прерывания
        self.write_reg8(self.SYSTEM_INTERRUPT_CLEAR, 0x01)

        print("✓ Датчик инициализирован")

    def start_ranging(self):
        """Запуск непрерывных измерений"""
        # Очистка прерывания
        self.write_reg8(self.SYSTEM_INTERRUPT_CLEAR, 0x01)

        # Запуск измерений (0x40 = ranging mode)
        self.write_reg8(self.SYSTEM_MODE_START, 0x40)

        # Небольшая задержка для первого измерения
        time.sleep(0.05)

    def stop_ranging(self):
        """Остановка измерений"""
        self.write_reg8(self.SYSTEM_MODE_START, 0x00)

    def get_distance(self):
        """
        Получение расстояния в миллиметрах

        Returns:
            Расстояние в мм или None при ошибке
        """
        try:
            # Проверяем готовность данных
            gpio_status = self.read_reg8(self.GPIO_TIO_HV_STATUS)

            if not (gpio_status & 0x01):
                return None  # Данные не готовы

            # Читаем статус измерения
            range_status = self.read_reg8(self.RESULT_RANGE_STATUS)

            # Читаем результат измерения
            distance = self.read_reg16(self.RESULT_FINAL_CROSSTALK_CORRECTED_RANGE_MM_SD0)

            # Очищаем прерывание для следующего измерения
            self.write_reg8(self.SYSTEM_INTERRUPT_CLEAR, 0x01)

            # Проверка валидности
            if distance > 0 and distance < 8000:
                # Проверяем статус (биты 4-7 должны быть 0 для валидного измерения)
                if (range_status >> 4) & 0x0F == 0:
                    return distance

            return None

        except Exception as e:
            return None

    def close(self):
        """Закрытие соединения"""
        if self.i2c_handle is not None:
            self.stop_ranging()
            print("✓ Датчик остановлен")


def main():
    print("=" * 50)
    print("VL53L1X Distance Sensor - WiringPi версия")
    print("=" * 50)

    # Создаем объект датчика
    sensor = VL53L1XWiringPi(i2c_bus=1, address=0x29)

    # Открываем соединение
    if not sensor.open():
        print("\n⚠ Проблемы с подключением датчика!")
        print("\nПроверьте:")
        print("1. Подключение:")
        print("   VIN → 3.3V (Pin 1)")
        print("   GND → GND (Pin 6)")
        print("   SCL → GPIO3/SCL (Pin 5)")
        print("   SDA → GPIO2/SDA (Pin 3)")
        print("\n2. Включен ли I2C:")
        print("   sudo raspi-config")
        print("   Interface Options → I2C → Enable")
        print("\n3. Установлена ли библиотека WiringPi:")
        print("   pip install wiringpi")
        sys.exit(1)

    # Инициализируем датчик
    sensor.init_sensor()

    # Запускаем измерения
    sensor.start_ranging()

    print("\nНачало измерений (Ctrl+C для остановки)...")
    print("-" * 50)

    count = 0
    errors = 0
    history = []
    last_distance = 0

    try:
        while True:
            distance = sensor.get_distance()

            if distance is not None:
                count += 1
                history.append(distance)

                # Храним последние 10 измерений
                if len(history) > 10:
                    history.pop(0)

                # Вычисляем среднее
                avg = sum(history) / len(history)

                # Изменение
                delta = distance - last_distance if last_distance > 0 else 0
                delta_str = f"Δ{delta:+4d}" if abs(delta) > 5 else "     "

                # Форматированный вывод
                timestamp = datetime.now().strftime("%H:%M:%S")
                print(f"[{timestamp}] #{count:4d}: {distance:4d} мм "
                      f"({distance/10:5.1f} см, {distance/1000:5.3f} м) "
                      f"Avg:{avg:6.1f} {delta_str}", end="")

                # Визуальная шкала
                bars = min(int(distance / 100), 40)
                print(" |" + "█" * bars)

                # Индикатор близости
                if distance < 100:
                    print("          ⚠️  ОЧЕНЬ БЛИЗКО!")
                elif distance < 300:
                    print("          ⚡ Близко")

                last_distance = distance
            else:
                errors += 1
                if errors % 20 == 0:
                    print(f"⚠ Ожидание данных... (пропусков: {errors})")

            # Частота опроса ~20 Гц
            time.sleep(0.05)

    except KeyboardInterrupt:
        print("\n" + "-" * 50)
        print("Статистика:")
        print(f"  Измерений: {count}")
        print(f"  Пропусков: {errors}")
        if count > 0:
            print(f"  Успешность: {count/(count+errors)*100:.1f}%")
            if history:
                print(f"  Последнее: {history[-1]} мм")
                print(f"  Среднее: {sum(history)/len(history):.1f} мм")
                print(f"  Мин/Макс: {min(history)}/{max(history)} мм")

    finally:
        sensor.close()
        print("\n✓ Программа завершена")


if __name__ == "__main__":
    main()

File: vl53l1x_wiringpi.c
/*
 * VL53L1X Distance Sensor - WiringPi C версия
 * Компилировать: gcc -o vl53l1x_wiringpi vl53l1x_wiringpi.c -lwiringPi
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <wiringPi.h>
#include <wiringPiI2C.h>

// I2C адрес датчика
#define VL53L1X_ADDRESS 0x29

// Основные регистры VL53L1X
#define SOFT_RESET                      0x0000
#define FIRMWARE_SYSTEM_STATUS          0x0010
#define IDENTIFICATION_MODEL_ID         0x010F
#define SYSTEM_INTERRUPT_CLEAR          0x0086
#define SYSTEM_MODE_START               0x0087
#define GPIO_TIO_HV_STATUS             0x0031
#define RESULT_RANGE_STATUS            0x0089
#define RESULT_DISTANCE                0x0096

// Регистры конфигурации
#define RANGE_CONFIG_VCSEL_PERIOD_A    0x0060
#define RANGE_CONFIG_VCSEL_PERIOD_B    0x0063
#define RANGE_CONFIG_TIMEOUT_MACROP_A  0x005E
#define RANGE_CONFIG_TIMEOUT_MACROP_B  0x0061
#define RANGE_CONFIG_VALID_PHASE_HIGH  0x0069
#define SYSTEM_INTERRUPT_CONFIG_GPIO   0x0046

// Дескриптор I2C устройства
int fd;

// Функция записи в 16-битный регистр (8-битное значение)
void writeReg8(int fd, unsigned short reg, unsigned char value) {
    // Записываем старший байт адреса
    wiringPiI2CWriteReg8(fd, (reg >> 8) & 0xFF, reg & 0xFF);
    // Записываем значение
    wiringPiI2CWrite(fd, value);
}

// Функция записи в 16-битный регистр (16-битное значение)
void writeReg16(int fd, unsigned short reg, unsigned short value) {
    // Записываем адрес
    wiringPiI2CWriteReg8(fd, (reg >> 8) & 0xFF, reg & 0xFF);
    // Записываем данные (big-endian)
    wiringPiI2CWrite(fd, (value >> 8) & 0xFF);
    wiringPiI2CWrite(fd, value & 0xFF);
}

// Функция чтения из 16-битного регистра (8-битное значение)
unsigned char readReg8(int fd, unsigned short reg) {
    // Записываем адрес регистра
    wiringPiI2CWriteReg8(fd, (reg >> 8) & 0xFF, reg & 0xFF);
    // Читаем данные
    return wiringPiI2CRead(fd) & 0xFF;
}

// Функция чтения из 16-битного регистра (16-битное значение)
unsigned short readReg16(int fd, unsigned short reg) {
    unsigned char high, low;
    // Записываем адрес регистра
    wiringPiI2CWriteReg8(fd, (reg >> 8) & 0xFF, reg & 0xFF);
    // Читаем 2 байта
    high = wiringPiI2CRead(fd) & 0xFF;
    low = wiringPiI2CRead(fd) & 0xFF;
    return (high << 8) | low;
}

// Инициализация датчика
int initSensor(int fd) {
    unsigned short model_id;
    int i;

    printf("Инициализация VL53L1X...\n");

    // Проверка ID датчика
    model_id = readReg16(fd, IDENTIFICATION_MODEL_ID);
    if (model_id == 0xEACC) {
        printf("✓ Датчик VL53L1X обнаружен (ID: 0x%04X)\n", model_id);
    } else {
        printf("✗ Неизвестный датчик (ID: 0x%04X)\n", model_id);
        return -1;
    }

    // Программный сброс
    writeReg8(fd, SOFT_RESET, 0x00);
    usleep(10000); // 10ms
    writeReg8(fd, SOFT_RESET, 0x01);
    usleep(10000); // 10ms

    // Ждем загрузки прошивки
    printf("Ожидание загрузки прошивки...\n");
    for (i = 0; i < 100; i++) {
        if (readReg8(fd, FIRMWARE_SYSTEM_STATUS) & 0x01) {
            printf("✓ Прошивка загружена (попытка %d)\n", i + 1);
            break;
        }
        usleep(10000); // 10ms
    }

    if (i >= 100) {
        printf("✗ Таймаут загрузки прошивки\n");
        return -1;
    }

    // Базовая конфигурация
    writeReg8(fd, RANGE_CONFIG_VCSEL_PERIOD_A, 0x09);
    writeReg8(fd, RANGE_CONFIG_VCSEL_PERIOD_B, 0x0D);
    writeReg8(fd, RANGE_CONFIG_VALID_PHASE_HIGH, 0xC8);

    // Timing budget
    writeReg16(fd, RANGE_CONFIG_TIMEOUT_MACROP_A, 0x00D6);
    writeReg16(fd, RANGE_CONFIG_TIMEOUT_MACROP_B, 0x00D6);

    // Настройка прерываний
    writeReg8(fd, SYSTEM_INTERRUPT_CONFIG_GPIO, 0x01);
    writeReg8(fd, SYSTEM_INTERRUPT_CLEAR, 0x01);

    printf("✓ Датчик инициализирован\n");
    return 0;
}

// Запуск измерений
void startRanging(int fd) {
    writeReg8(fd, SYSTEM_INTERRUPT_CLEAR, 0x01);
    writeReg8(fd, SYSTEM_MODE_START, 0x40);
    usleep(50000); // 50ms для первого измерения
}

// Остановка измерений
void stopRanging(int fd) {
    writeReg8(fd, SYSTEM_MODE_START, 0x00);
}

// Получение расстояния
int getDistance(int fd) {
    unsigned char gpio_status, range_status;
    unsigned short distance;

    // Проверяем готовность данных
    gpio_status = readReg8(fd, GPIO_TIO_HV_STATUS);
    if (!(gpio_status & 0x01)) {
        return -1; // Данные не готовы
    }

    // Читаем статус измерения
    range_status = readReg8(fd, RESULT_RANGE_STATUS);

    // Читаем результат
    distance = readReg16(fd, RESULT_DISTANCE);

    // Очищаем прерывание
    writeReg8(fd, SYSTEM_INTERRUPT_CLEAR, 0x01);

    // Проверка валидности
    if (distance > 0 && distance < 8000) {
        // Проверяем статус
        if (((range_status >> 4) & 0x0F) == 0) {
            return distance;
        }
    }

    return -1;
}

// Визуализация расстояния
void printDistance(int distance, int count) {
    time_t now;
    struct tm *tm_info;
    char time_str[20];
    int bars, i;

    // Получаем текущее время
    time(&now);
    tm_info = localtime(&now);
    strftime(time_str, 20, "%H:%M:%S", tm_info);

    // Выводим данные
    printf("[%s] #%4d: %4d мм (%.1f см, %.3f м) ",
           time_str, count, distance,
           distance / 10.0, distance / 1000.0);

    // Визуальная шкала
    bars = distance / 100;
    if (bars > 40) bars = 40;
    printf("|");
    for (i = 0; i < bars; i++) {
        printf("█");
    }

    // Индикаторы близости
    if (distance < 100) {
        printf(" ⚠️ ОЧЕНЬ БЛИЗКО!");
    } else if (distance < 300) {
        printf(" ⚡ Близко");
    }

    printf("\n");
    fflush(stdout);
}

int main() {
    int distance;
    int count = 0;
    int errors = 0;
    int sum = 0;
    int min_dist = 9999;
    int max_dist = 0;

    printf("==========================================\n");
    printf("VL53L1X Distance Sensor - WiringPi C\n");
    printf("==========================================\n\n");

    // Инициализация WiringPi
    if (wiringPiSetup() == -1) {
        printf("✗ Ошибка инициализации WiringPi\n");
        return 1;
    }

    // Открываем I2C устройство
    fd = wiringPiI2CSetup(VL53L1X_ADDRESS);
    if (fd == -1) {
        printf("✗ Не удалось открыть I2C устройство\n");
        printf("\nПроверьте:\n");
        printf("1. Подключение датчика:\n");
        printf("   VIN → 3.3V (Pin 1)\n");
        printf("   GND → GND (Pin 6)\n");
        printf("   SCL → GPIO3 (Pin 5)\n");
        printf("   SDA → GPIO2 (Pin 3)\n");
        printf("2. Включен ли I2C:\n");
        printf("   sudo raspi-config → Interface Options → I2C\n");
        printf("3. Запустите с правами root:\n");
        printf("   sudo ./vl53l1x_wiringpi\n");
        return 1;
    }

    printf("✓ I2C устройство открыто (fd: %d)\n", fd);

    // Инициализация датчика
    if (initSensor(fd) != 0) {
        printf("✗ Ошибка инициализации датчика\n");
        return 1;
    }

    // Запуск измерений
    startRanging(fd);

    printf("\nНачало измерений (Ctrl+C для остановки)...\n");
    printf("------------------------------------------\n");

    // Основной цикл измерений
    while (1) {
        distance = getDistance(fd);

        if (distance > 0) {
            count++;
            sum += distance;

            if (distance < min_dist) min_dist = distance;
            if (distance > max_dist) max_dist = distance;

            printDistance(distance, count);
        } else {
            errors++;
            if (errors % 20 == 0) {
                printf("⚠ Ожидание данных... (пропусков: %d)\n", errors);
            }
        }

        usleep(50000); // 50ms = 20Hz
    }

    // Завершение (не достигается из-за бесконечного цикла)
    stopRanging(fd);
    close(fd);

    return 0;
}